### PR TITLE
[GNA] Remove warning suppression, fix all warnings

### PR DESCRIPTION
**Pull request by rjeziers**
### Details:
Removed warning suppression from the plugin's CMake.txt. Fixed all warnings.
Most of the warnings were related to implicit type conversion, specifically `size_t` -> `uint32_t`, because of inconsistent usage of these types for sizes and indices (OpenVino core and new code in GNA plugin use mostly `size_t`; legacy GNA code and GNA api use mostly `uint32_t`).
I refactored some of the plugin code to use `size_t` for sizes where it wasn't problematic. In the remaining places, static casts were added.

### Tickets:
 - 93715

### CODE REVIEW

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp [hunk 1]
```diff
@@ -306,8 +306,8 @@ bool ScaleFactorCalculator::requantizeInput(InferenceEngine::CNNLayerPtr input,
         return false;
     }
 
-    auto prevLayer = InferenceEngine::CNNNetHasPrevLayer(layer.get(), prevInputIdx)
-                         ? InferenceEngine::CNNNetPrevLayer(layer, prevInputIdx)
+    auto prevLayer = InferenceEngine::CNNNetHasPrevLayer(layer.get(), static_cast<int>(prevInputIdx))
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:309]: you set second argument to size_t and not you cast to int? 
inline bool CNNNetHasPrevLayer(const InferenceEngine::CNNLayer* layer, size_t idx = 0) {

shouldn't this produce warning?

https://github.com/openvinotoolkit/openvino/pull/17841/files#diff-637b2179f395ecedb9fb41cb1ee70aa666b3b7c2ce3d8fb435b9c3c8bbc248ffR355

File: src/plugins/intel_gna/src/common/numerical_utils.hpp [hunk 2]
```diff
@@ -28,6 +28,18 @@ inline int32_t FloatToInt32(float a) {
 inline int64_t FloatToInt64(float a) {
     return FloatToInteger<int64_t>(a);
 }
+inline int8_t FloatToInt8(double a) {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/common/numerical_utils.hpp:31]:
```suggestion
//Option 1. Add method for doubles.
template <typename T>
inline T DoubleToInteger(Double a) {
    return static_cast<T>((a < 0.0f) ? (a - 0.5f) : (a + 0.5f));
}
//Option 2. or improve template methods.
template <typename T, Typename U,  typename std::enable_if<std::is_floating_point<U>::value>::type* = nullptr> 
inline T DoubleToInteger(U a) {
    return static_cast<T>((a < 0.0f) ? (a - static_cast<U>(0.5)) : (a + static_cast<U>(0.5)));
}
```

File: src/plugins/intel_gna/src/common/numerical_utils.hpp [hunk 3]
```diff
@@ -28,6 +28,18 @@ inline int32_t FloatToInt32(float a) {
...
+    return FloatToInteger<int8_t>(static_cast<float>(a));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/common/numerical_utils.hpp:32]: Rename to DoubleToInt16
and remove casting.

File: src/plugins/intel_gna/src/frontend/model_quantizer.hpp [hunk 4]
```diff
@@ -92,10 +92,10 @@ class ModelQuantizer {
             // We are looking for infinite loop by using algorithm of compute prefix function, complexity O(N)
             // (a part of the Knuth–Morris–Pratt algorithm).
             std::map<int, int> prefix_function;
-            int32_t k = inf_loop_history.size();
-            for (int32_t i = inf_loop_history.size() - 2; i >= 0; i--) {
+            size_t k = inf_loop_history.size();
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/model_quantizer.hpp:95]:
```suggestion
            auto k = inf_loop_history.size();
```

File: src/plugins/intel_gna/src/frontend/model_quantizer.hpp [hunk 5]
```diff
@@ -127,9 +127,9 @@ class ModelQuantizer {
                         log::debug() << "\t " << s << '\n';
                     }
                     inf_loop_pattern.clear();
-                    int pattern_len = (inf_loop_history.size() - i) / 2;
+                    size_t pattern_len = (inf_loop_history.size() - i) / 2;
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/model_quantizer.hpp:130]:
```suggestion
                    auto pattern_len = (inf_loop_history.size() - i) / 2;
```

File: src/plugins/intel_gna/src/frontend/model_quantizer.hpp [hunk 6]
```diff
@@ -141,7 +141,7 @@ class ModelQuantizer {
                     break;
                 }
 
-                prefix_function.emplace(i, k);
+                prefix_function.emplace(i, static_cast<int>(k));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/model_quantizer.hpp:144]: can prefix_function have negative values? if not maybe it would be good to change map parameters types to size_t/uint32_t?

File: src/plugins/intel_gna/src/frontend/quantization.hpp [hunk 7]
```diff
@@ -81,7 +81,7 @@ inline T SaturationCast(float value, uint32_t* saturation_counter = nullptr) {
 /**
  * @brief Apply FQ levels onto a value
  */
-float ApplyFQ(float value, float input_low, float input_high, float output_low, float output_high, uint32_t levels);
+float ApplyFQ(float value, float input_low, float input_high, float output_low, float output_high, size_t levels);
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/quantization.hpp:84]: do you need size_t to uint32_t because of some specific reason or because you changed argument type used for invocation of the method from auto to size_t?

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp [hunk 8]
```diff
@@ -261,7 +261,7 @@ bool ScaleFactorCalculator::requantizeInput(InferenceEngine::CNNLayerPtr input,
     if (info.isWeightableIdentity() && !AreFpEq(quantDataForInputLayer->_weights_quant.GetScale(), 1.0f)) {
         auto reducer = std::max(1.0f, quantDataForInputLayer->_dst_quant.GetScale() / newOutputScale);
         auto newWeightsScale = std::max(1.0f, quantDataForInputLayer->_weights_quant.GetScale() / reducer);
-        quantDataForInputLayer->_weights_quant.SetScale(static_cast<int32_t>(newWeightsScale));
+        quantDataForInputLayer->_weights_quant.SetScale(trunc(newWeightsScale));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:264]: don't you have warning here?
It looks like newWeightsScale is double and trunc returns double, and later you pass it to SetScale which takes float.

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp [hunk 9]
```diff
@@ -410,7 +410,8 @@ float ScaleFactorCalculator::getActivationScale(InferenceEngine::CNNLayer const*
 #endif
     } else if (layer.isRelu()) {
         // if activation is one from relu family, we need to apply heuristic to avoid activation output overflow
-        auto limit = (inputsSize == 1 ? std::numeric_limits<int8_t>::max() : std::numeric_limits<int32_t>::max()) - 1;
+        auto limit = static_cast<unsigned int>(
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:413]: it is later used to comapred with uint64_t why don't use this type for casting?

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp [hunk 10]
```diff
@@ -800,7 +801,7 @@ bool ScaleFactorCalculator::ScaleFactorPerLayerCNN(InferenceEngine::CNNLayer* cn
         }
         return true;
     } else if (layerInfo.isCropAffined()) {
-        auto weightsScaleFactor = 1;
+        auto weightsScaleFactor = 1.0f;
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:804]:
```suggestion
        const auto weightsScaleFactor = 1.0f;
```

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp [hunk 11]
```diff
@@ -921,7 +922,7 @@ bool ScaleFactorCalculator::ScaleFactorPerLayerEltwise(InferenceEngine::EltwiseL
             return true;
         }
 
-        auto weightsReducer = calculateWeightsReducerFromDstStats(quantData->_dst_quant);
+        float weightsReducer = static_cast<float>(calculateWeightsReducerFromDstStats(quantData->_dst_quant));
         if (weightsReducer > initial_weights_reducer_val) {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:926]: initial_weights_reducer_val is double and it was compared to double , but you changed it to float.
This is quite risky, due the fact precision is very important here. I would not change type from double to float.

If this is only to fix warning in line 927 I would change it like that.
```
            float newOutputScale = quantParams1->_dst_quant.GetScale() / weightsReducer;
//to
            float newOutputScale = quantParams1->_dst_quant.GetScale() / static_cast<float>(weightsReducer);

```

File: src/plugins/intel_gna/src/gna_device_allocation.hpp [hunk 12]
```diff
@@ -84,7 +84,7 @@ struct GnaAllocation {
     }
 
     uint32_t sizeForExport() const {
-        return ALIGN64(sizeRequested);
+        return ALIGN64(static_cast<uint32_t>(sizeRequested));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_device_allocation.hpp:87]: why just not change type of sizeRequested in header?

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 13]
```diff
@@ -337,15 +337,23 @@ void GNAGraphCompiler::ConvolutionPrimitive(InferenceEngine::CNNLayerPtr layer)
...
-    const auto in_channels = InferenceEngine::GetDataDimByName(inputs, InferenceEngine::DataDimName::C);
-    auto in_height = InferenceEngine::GetDataDimByName(inputs, InferenceEngine::DataDimName::H);
-    auto in_width = InferenceEngine::GetDataDimByName(inputs, InferenceEngine::DataDimName::W);
-
-    const auto out_batch = InferenceEngine::GetDataDimByName(outputs, InferenceEngine::DataDimName::N);
-    const auto out_channels = InferenceEngine::GetDataDimByName(outputs, InferenceEngine::DataDimName::C);
-    auto out_height = InferenceEngine::GetDataDimByName(outputs, InferenceEngine::DataDimName::H);
-    auto out_width = InferenceEngine::GetDataDimByName(outputs, InferenceEngine::DataDimName::W);
+    const uint32_t in_batch =
+        static_cast<uint32_t>(InferenceEngine::GetDataDimByName(inputs, InferenceEngine::DataDimName::N));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:341]: are this changes needed?

GetDataDimByName returns uint32_t
so you don't need change auto to uint32_t and you don't need to use static_cast.

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 14]
```diff
@@ -543,10 +551,10 @@ void GNAGraphCompiler::finalizeConvolution1DPrimitive(InferenceEngine::CNNLayerP
     auto biasPrecision =
         convolution._biases ? convolution._biases->getTensorDesc().getPrecision() : outputs->getPrecision();
 
-    uint32_t num_bytes_per_input = inputs->getPrecision().size();
-    uint32_t num_bytes_per_output = outputs->getPrecision().size();
-    uint32_t num_bytes_per_weight = convolution._weights->getTensorDesc().getPrecision().size();
-    uint32_t num_bytes_per_bias = biasPrecision.size();
+    uint32_t num_bytes_per_input = static_cast<uint32_t>(inputs->getPrecision().size());
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:554]: you have to change all this types because you changes size_t to uint32_t.

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 15]
```diff
@@ -1417,23 +1434,24 @@ void GNAGraphCompiler::EltwisePrimitive(InferenceEngine::CNNLayerPtr layer) {
...
-        quantized == nullptr ? inputs2Bytes->getPrecision().size() : (gna_config.gnaFlags.input_low_precision ? 1 : 2),
-        quantized == nullptr ? inputs4Bytes->getPrecision().size() : (gna_config.gnaFlags.input_low_precision ? 1 : 4),
-        GetScaleFactor(layer, QuantizedDataType::weights),
-        GetScaleFactor(layer, QuantizedDataType::output),
-        ptr_inputs,
-        ptr_outputs,
-        ptr_weights,
-        ptr_biases,
-        true);
+    dnn->InitAffineComponent(currentComponent,
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:1437]: was this not formatted properly earlier?

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 16]
```diff
@@ -1520,12 +1538,12 @@ void GNAGraphCompiler::GemmPrimitive(InferenceEngine::CNNLayerPtr layer) {
...
     auto in_dims = input_1->getDims();
-    auto batch_size = (in_dims.size() == 1) ? 1 : in_dims.front();
-    uint32_t num_rows_in = InferenceEngine::details::product(in_dims) / batch_size;
+    uint32_t batch_size = (in_dims.size() == 1) ? 1 : static_cast<uint32_t>(in_dims.front());
+    uint32_t num_rows_in = static_cast<uint32_t>(InferenceEngine::details::product(in_dims)) / batch_size;
     uint32_t num_columns_in = batch_size;
     const auto out_dims = outputs->getDims();
     const auto out_dims_size = ngraph::shape_size(out_dims);
-    uint32_t num_rows_out = InferenceEngine::GetDimFromBack(out_dims, 1);
+    uint32_t num_rows_out = static_cast<uint32_t>(InferenceEngine::GetDimFromBack(out_dims, 1));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:1546]: does not GetDimFromBack return uint32_t by default?

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 17]
```diff
@@ -2041,10 +2060,10 @@ void GNAGraphCompiler::PWLPrimitive(InferenceEngine::CNNLayerPtr layer) {
 
     auto orientation = kDnnInterleavedOrientation;
 
-    uint32_t w_dim_in = GetDataDimByName(inputs, DataDimName::W);
-    uint32_t h_dim_in = GetDataDimByName(inputs, DataDimName::H);
-    uint32_t c_dim_in = GetDataDimByName(inputs, DataDimName::C);
-    uint32_t n_dim_in = GetDataDimByName(inputs, DataDimName::N);
+    uint32_t w_dim_in = static_cast<uint32_t>(GetDataDimByName(inputs, DataDimName::W));
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:2063]: GetDataDimByName returns uint32_t by default

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp [hunk 18]
```diff
@@ -2588,8 +2607,8 @@ void GNAGraphCompiler::connectOutput(InferenceEngine::CNNLayerPtr layer, void* p
 ConnectionDetails GNAGraphCompiler::connectInput(CNNLayerPtr layer,
                                                  void* ptr,
                                                  size_t num_data_bytes_in,
-                                                 int32_t offset,
-                                                 int idx,
+                                                 size_t offset,
+                                                 size_t idx,
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_compiler.cpp:2611]: when you change `int` to `size_t` ensure that it is safe in many places in code negative values is used to find issue or special case - in such case you have unexpected behavior if you lost the sign.

File: src/plugins/intel_gna/src/gna_graph_tools.hpp [hunk 19]
```diff
@@ -86,7 +86,7 @@ inline InferenceEngine::CNNLayerPtr CNNNetPrevLayer(const InferenceEngine::CNNLa
  * @param idx - index in previous layer collection
  * @param layer
  */
-inline InferenceEngine::CNNLayerPtr CNNNetPrevLayer(const InferenceEngine::CNNLayer* layer, int idx = 0) {
+inline InferenceEngine::CNNLayerPtr CNNNetPrevLayer(const InferenceEngine::CNNLayer* layer, size_t idx = 0) {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_tools.hpp:89]: ensure that negative value is not used as condition before you change int to size_t

**Review by marcinkusm — CHANGES_REQUESTED** [review]

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp [hunk 20]
```diff
@@ -284,7 +284,7 @@ bool ScaleFactorCalculator::requantizeInput(InferenceEngine::CNNLayerPtr input,
                 break;
             }
         }
-        auto prevLayer = InferenceEngine::CNNNetPrevLayer(layer, prevInputIdx);
+        auto prevLayer = InferenceEngine::CNNNetPrevLayer(layer, static_cast<int>(prevInputIdx));
```
[the remark this answers is not in the document]
Comment (rjeziers) [reply src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:287]: A leftover from before I changed the function signatures - I'll check all places

File: src/plugins/intel_gna/src/gna_graph_compiler.cpp (hunk 15 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/gna_graph_compiler.cpp:1437]: I think formatting changed due to additional casting of the arguments

File: src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp (hunk 8 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/frontend/scale_factor_calc.cpp:264]: newWeightsScale is float, so trunc also returns a float.

File: src/plugins/intel_gna/src/frontend/model_quantizer.hpp (hunk 6 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/frontend/model_quantizer.hpp]: `prefix_function` can't have negative values, but `i` can, in fact `i >= 0` is a condition for searching an infinite loop; i've changed map parameter types to <int, size_t>

File: src/plugins/intel_gna/src/frontend/model_quantizer.hpp [hunk 21]
```diff
@@ -57,7 +57,7 @@ class ModelQuantizer {
             auto input_layer = getCreatorLayer(inputData.second->getInputData()).lock();
             auto quant_data = InferenceEngine::getInjectedData<frontend::QuantizedLayerParams>(input_layer);
             IE_ASSERT(quant_data != nullptr);
-            quant_data->_src_quant.SetScale(inputs.at(input_layer->name).scale_factor);
+            quant_data->_src_quant.SetScale(static_cast<float>(inputs.at(input_layer->name).scale_factor));
```
Comment (tadamowicz) [inline src/plugins/intel_gna/src/frontend/model_quantizer.hpp:60]: I would suggest change type of the scale_factor from double to float

File: src/plugins/intel_gna/src/optimizer/gna_pass_manager.cpp [hunk 22]
```diff
@@ -486,22 +486,23 @@ void ReorderMaxPoolPass::run() {
...
+     * power |                  power |
+     *  \   /                    \   /
+     *    mul                      mul
+     */
 
     auto hasNChildren = [](CNNLayerPtr l, int N) {
         if (l->outData.size() != 1)
             return false;
-        if (getInputTo(l->outData.front()).size() != N)
+        if (static_cast<int>(getInputTo(l->outData.front()).size()) != N)
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/optimizer/gna_pass_manager.cpp:505]: just wonder if it wouldn't be better to make N as size_t

File: src/plugins/intel_gna/src/gna_plugin.cpp [hunk 23]
```diff
@@ -1347,11 +1348,11 @@ RequestStatus GNAPlugin::WaitFor(uint32_t request_idx, int64_t millisTimeout) {
         ExportScores(output_blob->buffer(),
                      gna_output_blob->cbuffer(),
                      gna_output_desc.orientation,
-                     batchSize,
-                     batchSize,
-                     elementsPerBatch,
-                     elementsPerBatch,
-                     elementsPerBatch,
+                     static_cast<uint32_t>(batchSize),
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_plugin.cpp:1351]: what is a type of batchSize and elementsPerBatch?
cannot they be declared as uint32_t to avoid casting in many places?

File: src/plugins/intel_gna/src/runtime/floatmath.cpp [hunk 24]
```diff
@@ -9,10 +9,6 @@
 #include <cstdint>
 #include <cstdio>
 
-#ifdef __cplusplus
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/runtime/floatmath.cpp:12]: are you sure about this removal ?

File: src/plugins/intel_gna/src/serial/gna_model_serial.cpp [hunk 25]
```diff
@@ -63,7 +63,7 @@ inline void write_pre_processing_model(const std::shared_ptr<ov::Model>& model,
     writeString(xml_buf.str(), os);
 
     // write BIN
-    size_t ir_bin_size = bin_buf.str().size();
+    uint32_t ir_bin_size = static_cast<uint32_t>(bin_buf.str().size());
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/serial/gna_model_serial.cpp:66]: please check writeBits it is template method and it uses size of template type for writing a file of model.
you cannot change a type passed there.

File: src/plugins/intel_gna/src/serial/gna_model_serial.cpp [hunk 26]
```diff
@@ -44,8 +44,8 @@ inline void writeBits(const T& obj, std::ostream& os) {
 
 inline void writeString(const std::string& str, std::ostream& os) {
     const char* c_str = str.c_str();
-    const size_t str_len = strlen(c_str) + 1;
-    writeBits(static_cast<uint32_t>(str_len), os);
+    const uint32_t str_len = static_cast<uint32_t>(strlen(c_str)) + 1;
+    writeBits(str_len, os);
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/serial/gna_model_serial.cpp:48]: don't change a type of variable passes to writeBits or writeNBytes

File: src/plugins/intel_gna/src/backend/am_intel_dnn.cpp [hunk 27]
```diff
@@ -1386,7 +1386,7 @@ void AMIntelDNN::InitGNAStruct(Gna2Model* gnaModel) {
         THROW_GNA_EXCEPTION << "out of memory in AMIntelDNN::InitGNAStruct()";
     memset(gnaModel->Operations, 0, gnaModel->NumberOfOperations * sizeof(Gna2Operation));
     gnaOperation = gnaModel->Operations;
-    for (int i = 0; i < component.size(); i++) {
+    for (int i = 0; i < static_cast<int>(component.size()); i++) {
```
[the remark this answers is not in the document]
Comment (rjeziers) [reply src/plugins/intel_gna/src/backend/am_intel_dnn.cpp:1389]: `i` is used as an argument in 7 different places - if it was `size_t` we'd need to cast it to `int` 7 times, this is a better option imho

File: src/plugins/intel_gna/src/backend/gna_limitations.cpp [hunk 28]
```diff
@@ -944,7 +945,7 @@ bool Limitations::validate_concat_axis(const InferenceEngine::CNNLayerPtr layer,
                        ptr->GetParamAsInts("order") == std::vector<int32_t>{0, 2, 1} /* NCW to NWC */))));
         };
 
-        for (auto input_idx = 0; input_idx != concat_layer->insData.size(); input_idx++) {
+        for (auto input_idx = 0; input_idx != static_cast<int>(concat_layer->insData.size()); input_idx++) {
             prev_layer = InferenceEngine::CNNNetPrevLayerSkipCertain(layer, input_idx, isFusableWithConv);
```
[the remark this answers is not in the document]
Comment (rjeziers) [reply src/plugins/intel_gna/src/backend/gna_limitations.cpp:949]: With option 1, I'd need to:
1. cast this argument inside CNNNetPrevLayerSkipCertain again to int, or
2. refactor whole graph tools - something which I've tried to do before.

If we don't want to introduce a lot of refactoring in this PR, option 2 is better.

File: src/plugins/intel_gna/src/runtime/floatmath.cpp (hunk 24 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/runtime/floatmath.cpp:12]: For me, it looks like this file was copy-pasted from some library in the early stages of this project and nobody bothered to modify it afterwards. I don't think these functions are actually used anywhere else.

File: src/plugins/intel_gna/src/serial/gna_model_serial.cpp (hunk 26 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/serial/gna_model_serial.cpp:48]: I didn't change the type - before, str_len was size_t but it was casted to uint32_t as an argument to writeBits. Now it's declared as uint32_t to avoid unnecessary cast to uint32_t in the next line.

writeBits was and is called with uint32_t; writeNBytes accepts only uint32_t, so it was called with size_t with implicit conversion to uint32_t; now it's called explicitly with uint32_t.

Nothing changed here.

File: src/plugins/intel_gna/src/runtime/floatmath.cpp (hunk 24 above)
Comment (marcinkusm) [reply src/plugins/intel_gna/src/runtime/floatmath.cpp:12]: this is ifdef and extern C are used to provide C like api( to prevent name mangling).
If you are sure that it is not needed we can remove it.

File: src/plugins/intel_gna/src/backend/am_intel_dnn.hpp [hunk 29]
```diff
@@ -272,22 +272,22 @@ class AMIntelDNN {
     }
 
     template <class T>
-    void AdvanceOperationIfAllApplied(const std::vector<intel_dnn_component_t>& cmp, int i, T*& operation) {
+    void AdvanceOperationIfAllApplied(const std::vector<intel_dnn_component_t>& cmp, size_t i, T*& operation) {
         if (i == cmp.size() - 1 || cmp[i + 1].operation != kDnnPiecewiselinearOp) {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/backend/am_intel_dnn.hpp:276]: I suggested change to `i + 1 = cma.size()` becayse I wanted to avoid situation that cmp.size() will be 0 so substracting 1 would be undefined behaviour, but I understand it is not possible here.

File: src/plugins/intel_gna/src/gna_graph_tools.hpp [hunk 30]
```diff
@@ -173,12 +173,12 @@ inline std::pair<InferenceEngine::CNNLayerPtr, std::vector<int>> CNNNetCheckNext
     int iidx,
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/gna_graph_tools.hpp:173]: oidx, iidx, new_oidx  and new_iidx are compared everywhere to size_t  from  .size() of containers.
it they cannot have negavative value maybe it would be good to change their type to size_t?

File: src/plugins/intel_gna/src/serial/gna_model_serial.cpp [hunk 31]
```diff
@@ -149,7 +149,7 @@ header_latest::ModelHeader GNAModelSerial::ReadHeader(std::istream& is) {
     auto size_of_headers_header = sizeof(header_latest::ModelHeader::gnam) +
                                   sizeof(header_latest::ModelHeader::headerSize) +
                                   sizeof(header_latest::ModelHeader::Version);
-    if (stream_len > size_of_headers_header) {
+    if (static_cast<uint32_t>(stream_len) > size_of_headers_header) {
```
Comment (marcinkusm) [inline src/plugins/intel_gna/src/serial/gna_model_serial.cpp:152]: are you sure that stream_len at this point will be >= 0?

Comment (rjeziers) [reply src/plugins/intel_gna/src/serial/gna_model_serial.cpp:152]: yes, in order for it to be <0 the stream starting position would have to be greater than stream length which is impossible

File: src/plugins/intel_gna/src/backend/am_intel_dnn.hpp (hunk 29 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/backend/am_intel_dnn.hpp:276]: If `cmp.size() == 0` we'd have an out-of-bounds error anyway, in the next condition. I could add some check for correctness of the data, but:
- it has worked before
- it's legacy code anyway

do you think it's in scope of this task?

File: src/plugins/intel_gna/src/gna_graph_tools.hpp (hunk 30 above)
Comment (rjeziers) [reply src/plugins/intel_gna/src/gna_graph_tools.hpp:173]: oidx, iidx are arguments to the function, so I'd need to refactor a lot of other functions from gna_graph_tools.hpp, which is what we wanted to avoid in this PR - maybe something for the future

**Review by marcinkusm — APPROVED** [review]
👍
